src/cls/rbd/cls_rbd.cc

```diff
@@ -2652,9 +2652,9 @@ int metadata_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
   map<string, bufferlist> data;
   string last_read = metadata_key_for_name(start_after);
   int max_read = max_return ? MIN(RBD_MAX_KEYS_READ, max_return) : RBD_MAX_KEYS_READ;
-  bool more;
+  bool more = true;
 
-  do {
+  while (more && data.size() < max_return) {
```

Comment:
if `max_return` is `0`, then this loop will return immediately. but what we expect is: it will read `max_read` before returns.

Reply:
Hopefully addressed above by setting ```max_return``` to ```RBD_MAX_KEYS_READ``` if zero.